code/mob/living/critter/ai/dog.dm

```diff
@@ -0,0 +1,108 @@
+/datum/aiHolder/dog
+	New()
+		..()
+		default_task = get_instance(/datum/aiTask/prioritizer/critter/dog, list(src))
+
+/datum/aiTask/prioritizer/critter/dog/New()
+	..()
+	transition_tasks += holder.get_instance(/datum/aiTask/timed/wander, list(holder, src))
+
+/// Go fetch!
+/datum/aiTask/sequence/goalbased/critter/dog/fetch
+	name = "fetching"
+	weight = 10
+	max_dist = 7
+	ai_turbo = TRUE
+
+/datum/aiTask/sequence/goalbased/critter/dog/fetch/New(parentHolder, transTask)
+	..(parentHolder, transTask)
+	add_task(holder.get_instance(/datum/aiTask/succeedable/critter/dog/fetch, list(holder)))
+
+/datum/aiTask/sequence/goalbased/critter/dog/fetch/get_targets()
+	var/mob/living/critter/small_animal/dog/the_dog = holder.owner
+	return list(the_dog.fetch_item)
+
+//Fetch subtask, pick up the item
+/datum/aiTask/succeedable/critter/dog/fetch
+	name = "fetch subtask"
+	var/is_complete = FALSE
+
+/datum/aiTask/succeedable/critter/dog/fetch/failed()
+	var/mob/living/critter/C = holder.owner
+	var/obj/item/I = holder.target
+	if(!C || !I || BOUNDS_DIST(I, C) > 0) //the tasks fails and is re-evaluated if the target is not in range
```

Comment:
```suggestion
	if(!C || !I || BOUNDS_DIST(I, C) > 0 || !istype(I.loc, /turf)) //the tasks fails and is re-evaluated if the target is not in range
```
in case someone picks up the thrown object, which would end up with the dog stuck in this task forever